### ISSUE TITLE
archive.Import: too many open files

### ISSUE THREAD

**Issue by philrz** [brimsec/zq#1167]
Repro was on macOS with `zar` & `zq` commit `b7023d5`.

I happened to try importing the full 70 GB wrccdc set of Zeek logs. The failure message:

```
$ zq * | zar import -s 25MB -
archive.Import: run failed: open /var/folders/yn/jbkxxkpd4vg142pc3_bd_krc0000gn/T/zq-sort-293329507/248: too many open files
```

This class of problem has resurfaced multiple times in the iterations of software we've written in the past couple years. It's known that the default `ulimit -n` on macOS (or, at least, on whatever my Mac laptop happens to be running) is small compared to what we're accustomed to seeing on Linux systems:

```
$ ulimit -n
256
```

In these situations, the fallback plan has always been that we could just require users to up this setting before they can expect our software to work. However, in the past we've always successfully managed to make our software behave in a way that it doesn't _require_ as many open files to function/perform, and hence has a higher chance of "just working" out-of-the-box on all of our target platforms. Therefore I've opened this issue to see if we can repeat the same yet again.

**Comment by alfred-landrum** [comment]
From the error string, this happens during the time sorting of data that archive.Import uses.

**Comment by philrz** [comment]
Verified in `zq` commit `b2c8356`.

Following the original repro steps, my `zq * | zar import -s 25MB -` of that 70 GB test data set did now run to completion without error. As the import was happening, I watched in another window as the temp space for the disk-based merge sort grew in size and eventually flushed out. During several repros of the failure case, the `too many open files` error would consistently show up right as the temp directory got to 31 GB in size.

```
$ while true; do du -sh zq-sort-607514208/ ; sleep 60; done
394M	zq-sort-607514208/
1.1G	zq-sort-607514208/
1.8G	zq-sort-607514208/
2.4G	zq-sort-607514208/
3.2G	zq-sort-607514208/
4.1G	zq-sort-607514208/
5.0G	zq-sort-607514208/
5.6G	zq-sort-607514208/
6.1G	zq-sort-607514208/
6.8G	zq-sort-607514208/
7.5G	zq-sort-607514208/
8.3G	zq-sort-607514208/
8.9G	zq-sort-607514208/
9.6G	zq-sort-607514208/
 10G	zq-sort-607514208/
 11G	zq-sort-607514208/
 12G	zq-sort-607514208/
 12G	zq-sort-607514208/
 13G	zq-sort-607514208/
 13G	zq-sort-607514208/
 14G	zq-sort-607514208/
 15G	zq-sort-607514208/
 16G	zq-sort-607514208/
 16G	zq-sort-607514208/
 17G	zq-sort-607514208/
 18G	zq-sort-607514208/
 19G	zq-sort-607514208/
 19G	zq-sort-607514208/
 20G	zq-sort-607514208/
 21G	zq-sort-607514208/
 21G	zq-sort-607514208/
 22G	zq-sort-607514208/
 22G	zq-sort-607514208/
 23G	zq-sort-607514208/
 24G	zq-sort-607514208/
 24G	zq-sort-607514208/
 25G	zq-sort-607514208/
 26G	zq-sort-607514208/
 26G	zq-sort-607514208/
 27G	zq-sort-607514208/
 28G	zq-sort-607514208/
 28G	zq-sort-607514208/
 29G	zq-sort-607514208/
 30G	zq-sort-607514208/
 30G	zq-sort-607514208/
 31G	zq-sort-607514208/
 32G	zq-sort-607514208/
 33G	zq-sort-607514208/
 34G	zq-sort-607514208/
 34G	zq-sort-607514208/
 35G	zq-sort-607514208/
 36G	zq-sort-607514208/
 36G	zq-sort-607514208/
 37G	zq-sort-607514208/
 38G	zq-sort-607514208/
 38G	zq-sort-607514208/
 39G	zq-sort-607514208/
 40G	zq-sort-607514208/
 41G	zq-sort-607514208/
 41G	zq-sort-607514208/
 42G	zq-sort-607514208/
 43G	zq-sort-607514208/
 44G	zq-sort-607514208/
 44G	zq-sort-607514208/
 46G	zq-sort-607514208/
 47G	zq-sort-607514208/
 47G	zq-sort-607514208/
 48G	zq-sort-607514208/
 48G	zq-sort-607514208/
 49G	zq-sort-607514208/
 50G	zq-sort-607514208/
 48G	zq-sort-607514208/
 38G	zq-sort-607514208/
 31G	zq-sort-607514208/
 24G	zq-sort-607514208/
 18G	zq-sort-607514208/
9.5G	zq-sort-607514208/
3.1G	zq-sort-607514208/
du: zq-sort-607514208/: No such file or directory
```

I also have to report that I burned some time at first thinking the fix had not been effective, and I can't explain precisely what I saw, but I'll describe it here in the interest of good science. It started from the fact that when I sat down to verify the issue, the number that came back from `ulimit -n` on my Mac was now `2560` where it had been `256` when I first opened up the issue. This was true even in newly-opened shells where I was attempting to do the verification. I am not aware of having done anything to my Mac's config that would have changed this setting. The one low-level thing that came to mind that happened in the time since I opened this issue was that I'd accepted an upgrade to a more recent version of xcode. 🤷 

In any case, in that state, it seemed like the fix was not effective. If I left it at that higher setting of `2560`, the `zar import` ran fine both at the original repro commit `b7023d5` and the commit `b2c8356` that has the fix. So, FWIW, if my Mac had flipped in this way in the past, I'd never have seen the symptom that led me to file this issue! But what's even more strange is that in that state, if I manually set `ulimit -n 256` to try and make my system look like how it was back when I first filed the issue, _even when testing with commit `b2c8356` that has the fix_, I could trigger the `too many open files` error.

I was super stumped at that point, so like any engineer, at this point I rebooted my laptop. At that point, `ulimit -n` was returning `256` like back in the day. And in _that_ state I could consistently:

1. Repro the original `too many open files` symptom on commit `b7023d5` where I first saw the problem
2. No longer saw the `too many open files` symptom on commit `b2c8356` that has the fix

In conclusion, the fix appeared effective in the end, but I can't explain what I saw on that `2560` tangent. In the end after that reboot things seem sane, so... moving forward!

Thanks @nwt!
